File: duckdata/examples/simple_pipeline/scripts/analyze_data.jl
"""
Analyze data for the PipeLink example pipeline.

This is the third node in the pipeline that analyzes the transformed data
using CrossLink for zero-copy data sharing.
"""

using Statistics
using DataFrames
using JSON
using Dates
using CSV
using UUIDs  # Add UUID for consistent dataset IDs

# Import CrossLink with a more robust path resolution
function import_crosslink()
    try
        # First try standard path from setup_julia.jl
        crosslink_path = joinpath(dirname(Base.find_package("DuckDB")), "..", "..", "crosslink", "crosslink.jl")
        if isfile(crosslink_path)
            include(crosslink_path)
            return true
        end
    catch
        # Fallback options
    end
    
    try
        # Try direct path relative to script location
        crosslink_path = joinpath(dirname(@__DIR__), "..", "..", "pipelink", "crosslink", "julia", "crosslink.jl")
        if isfile(crosslink_path)
            include(crosslink_path)
            return true
        end
    catch
        # Fallback options
    end
    
    try
        # Try original path
        crosslink_path = joinpath(dirname(Base.find_package("DuckDB")), "..", "..", "..", "pipelink", "crosslink", "julia", "crosslink.jl")
        if isfile(crosslink_path)
            include(crosslink_path)
            return true
        end
    catch
        # Fallback options
    end
    
    return false
end

# Import CrossLink
if !import_crosslink()
    error("Failed to import CrossLink module. Please run setup_julia.jl first.")
end
using .CrossLink

println("Starting analyze_data.jl")

"""
    analyze_data(data, detailed=false)

Perform statistical analysis on the input data.
"""
function analyze_data(data, detailed=false)
    # Basic statistics
    stats = Dict{String, Any}()
    
    # Summary statistics for numeric columns
    numeric_cols = filter(col -> eltype(data[!, col]) <: Number, names(data))
    
    for col in numeric_cols
        if col != "id"  # Skip id column
            col_stats = Dict{String, Any}(
                "mean" => mean(data[!, col]),
                "median" => median(data[!, col]),
                "std" => std(data[!, col]),
                "min" => minimum(data[!, col]),
                "max" => maximum(data[!, col])
            )
            
            stats[string(col)] = col_stats
        end
    end
    
    # Convert to a flat DataFrame for storage
    result_rows = []
    
    for (key, value) in stats
        if value isa Dict
            for (subkey, subvalue) in value
                push!(result_rows, (statistic = "$(key)_$(subkey)", value = subvalue))
            end
        else
            push!(result_rows, (statistic = key, value = value))
        end
    end
    
    return DataFrame(result_rows)
end

# Main function to process node
function main()
    # Check command line arguments
    if length(ARGS) < 1
        error("No metadata file provided. This script should be run by PipeLink.")
    end
    
    # Load metadata from file
    meta_path = ARGS[1]
    println("Using metadata file: $meta_path")
    
    # Parse the metadata file to extract db_path
    meta_content = read(meta_path, String)
    db_path_line = match(r"db_path:\s*(.+)", meta_content)
    db_path = String(strip(db_path_line.captures[1]))
    println("Using db_path: $db_path")
    
    # Initialize CrossLink manager with the database
    println("Initializing CrossLink with database: $db_path")
    manager = CrossLinkManager(db_path, true)  # Enable debug mode
    
    # Get transformed data using CrossLink's pull_data
    println("Getting transformed data via CrossLink...")
    
    # Try with error handling
    try
        transformed_data = pull_data(manager, "transformed_data", zero_copy=true)
        
        println("Loaded data: $(nrow(transformed_data)) rows, columns: $(join(names(transformed_data), ", "))")
        
        # Analyze the data
        println("Analyzing data...")
        result = analyze_data(transformed_data)
        println("Analysis complete with $(nrow(result)) statistics calculated")
        
        # Save analysis results using CrossLink's push_data
        println("Saving analysis results with CrossLink...")
        try
            # Use push_data with memory mapping for zero-copy
            dataset_id = push_data(
                manager,
                result,
                "analysis_results",
                description="Statistical analysis of the transformed data",
                enable_zero_copy=true,
                memory_mapped=true,
                shared_memory=false,
                access_languages=["python", "r", "julia", "cpp"]
            )
            
            println("Analysis results saved with dataset ID: $dataset_id")
        catch e
            println("Error saving results: $e")
            # Print stack trace for debugging
            Base.showerror(stderr, e, catch_backtrace())
            rethrow(e)
        end
    catch e
        println("Error retrieving or processing data: $e")
        # Print stack trace for debugging
        Base.showerror(stderr, e, catch_backtrace())
        rethrow(e)
    end
end

# Run the main function
try
    main()
    println("analyze_data.jl completed")
catch e
    println("Error in analyze_data.jl: $e")
    Base.showerror(stderr, e, catch_backtrace())
    # Re-throw to ensure the pipeline knows there was an error
    rethrow(e)
end 

File: duckdata/examples/simple_pipeline/scripts/python/analyze_data.py
"""
Analyze data for the PipeLink example pipeline (Python version).

This is the Python equivalent of analyze_data.jl for performance comparison.
It analyzes the transformed data using CrossLink for zero-copy data sharing.
"""

import os
import sys
import json
import time
import numpy as np
import pandas as pd
import re
from datetime import datetime

# Import CrossLink (make sure it's in the PYTHONPATH)
sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), 
                            "../../../../pipelink"))
from crosslink.python.crosslink import CrossLink

print("Starting analyze_data.py")
start_time = time.time()

def analyze_data(data, detailed=False):
    """
    Perform statistical analysis on the input data.
    
    Args:
        data: Pandas DataFrame to analyze
        detailed: Whether to perform detailed analysis
        
    Returns:
        DataFrame with analysis results
    """
    # Basic statistics
    stats = {}
    
    # Get numeric columns
    numeric_cols = data.select_dtypes(include=['number']).columns.tolist()
    
    for col in numeric_cols:
        if col != 'id':  # Skip id column
            col_stats = {
                'mean': data[col].mean(),
                'median': data[col].median(),
                'std': data[col].std(),
                'min': data[col].min(),
                'max': data[col].max()
            }
            
            stats[str(col)] = col_stats
    
    # Convert to a flat DataFrame for storage
    result_rows = []
    
    for key, value in stats.items():
        if isinstance(value, dict):
            for subkey, subvalue in value.items():
                result_rows.append({
                    'metric': f"{key}_{subkey}", 
                    'value': subvalue
                })
        else:
            result_rows.append({
                'metric': key, 
                'value': value
            })
    
    return pd.DataFrame(result_rows)

def main():
    # Check command line arguments
    if len(sys.argv) < 2:
        raise ValueError("No metadata file provided. This script should be run by PipeLink.")
    
    # Load metadata from file
    meta_path = sys.argv[1]
    print(f"Using metadata file: {meta_path}")
    
    # Parse the metadata file to extract db_path
    with open(meta_path, 'r') as f:
        meta_content = f.read()
    
    db_path_match = re.search(r'db_path:\s*(.+)', meta_content)
    if not db_path_match:
        raise ValueError("Could not find db_path in metadata file")
    
    db_path = db_path_match.group(1).strip()
    print(f"Using db_path: {db_path}")
    
    # Initialize CrossLink with the database
    print("Initializing CrossLink with database...")
    crosslink_init_start = time.time()
    cl = CrossLink(db_path)
    crosslink_init_time = time.time() - crosslink_init_start
    print(f"CrossLink initialized in {crosslink_init_time:.6f} seconds")
    
    # Get transformed data using zero-copy
    print("Getting transformed data via CrossLink (zero-copy)...")
    pull_start = time.time()
    transformed_data = cl.pull("transformed_data", zero_copy=True)
    pull_time = time.time() - pull_start
    print(f"Loaded data via zero-copy in {pull_time:.6f} seconds: {len(transformed_data)} rows, columns: {', '.join(transformed_data.columns)}")
    
    # Analyze the data
    print("Analyzing data...")
    analysis_start = time.time()
    result = analyze_data(transformed_data)
    analysis_time = time.time() - analysis_start
    print(f"Analysis complete with {len(result)} statistics calculated in {analysis_time:.6f} seconds")
    
    # Save results using CrossLink's push method
    print("Saving analysis results with CrossLink...")
    push_start = time.time()
    dataset_id = cl.push(
        result, 
        "analysis_results",
        description="Statistical analysis of the transformed data (Python version)",
        use_arrow=True
    )
    push_time = time.time() - push_start
    print(f"Analysis results shared via CrossLink with ID: {dataset_id} in {push_time:.6f} seconds")
    
    # Total time
    total_time = time.time() - start_time
    print(f"Total execution time: {total_time:.6f} seconds")
    
    # Write timing information to a file
    timing_info = {
        "language": "python",
        "crosslink_init_time": crosslink_init_time,
        "data_pull_time": pull_time,
        "analysis_time": analysis_time,
        "data_push_time": push_time,
        "total_time": total_time,
        "rows_processed": len(transformed_data),
        "columns_processed": len(transformed_data.columns),
        "timestamp": datetime.now().isoformat()
    }
    
    with open("python_timing.json", "w") as f:
        json.dump(timing_info, f, indent=2)

if __name__ == "__main__":
    main()
    print("analyze_data.py completed") 